docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,616 @@
+will contain hundreds of individual passes. Often external users will want to
+have custom passes correctly scheduled without having to modify a single
+handcrafted pass order.
+
+On the other hand, modern deep learning frameworks, such as Pytorch and MXNet
+Gluon, also have the tendency to enable pass style layer construction
+scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
+these modern frameworks are able to conveniently add modules/layers to their
+containers and build up neural network easily.
+
+The design of the Relay pass infra is largely inspired by the the hierarchical
+pass manager used in LLVM and the block style containers used in the popular
+deep learning frameworks. The major goals of the pass infra include:
+
+#) enabling better programmatic orchestration of optimizations. This allows
+   users to flexibly customize and build their own optimization pipelines.
+
+#) providing a user friendly way to debug optimization passes.
+
+#) alleviating developers from manually and respectively resolving the
+   dependencies between passes.
+
+#) simplifying the implementation of new passes for developers. For example, we
+   allow users to implement a pass in Python and let the pass infra manipulate
+   its execution.
+
+The Design
+----------
+
+We focus on ease of extension for users, making it possible for users to quickly
+add new passes without loss of backward compatibility. The design contains both
+the backend and the frontend. The former implements the main logic of the pass
+infra. The latter provides simple APIs for users to interact with, i.e.
+allowing users to quickly create their own optimization pipelines.
+
+C++ Backend
+~~~~~~~~~~~
+
+We provide a ``PassInfo`` object to contain the basic information needed by
+a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization
+level the pass will be enabled, and ``required`` represents the passes that are
+required to execute a certain pass (see `include/tvm/relay/transform.h`_ for
+more details).
+
+.. code:: c++
+
+    class PassInfoNode : public RelayNode {
+      std::string name;
+      int opt_level;
+      std::vector<std::string> required;
+    };
+
+PassContext
+^^^^^^^^^^^
+
+``PassContext`` carries useful information for an optimization pass. For
+example, it contains the error reporting system so optimization authors can
+provide diagnostics about why an optimization fails. ``PassContext`` is also
+designed to replace the old ``BuildConfig`` which was used to help users
+configure the compilation options, including optimization level and
+required/disabled passes, etc. For instance, we may have a configuration which
+performs all passes at ``opt_level=3`` with some disabled passes using
+``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes
+at ``opt_level=3`` and exclude those in disabled pass list.
+
+This class is designed as a thread local so that users can conveniently use the
```

Comment:
What does "thread-local" mean in this context?

Reply:
I will rephrase this. I wanted to say that we used a thread-local store to keep the pass context objects. Users can get the current context in a thread-safe way.